JIT pass for add relu fusion.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40252 Add add_relu fusion pass to optimize_for_mobile.
* **#39343 JIT pass for add relu fusion.**
* #39342 Add fused add_relu op.
* #39341 [Vec256][neon] Add neon backend for vec256
* #40059 Add benchmark for add op.

Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

kimishpatel committed:
JIT pass for add relu fusion.

Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass for add relu fusion."


Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass for add relu fusion."


Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass for add relu fusion."


Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass for add relu fusion."


Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass for add relu fusion."


Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass for add relu fusion."


Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass for add relu fusion."


Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass for add relu fusion."


Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

[ghstack-poisoned]

dreiss commented on test/test_jit.py:
```diff
@@ -545,6 +545,153 @@ def f(x, y):
+                x = self.relu_op(x)
+                return x
+        a = torch.rand((7, 11))
+        a = a * -10
+        a = a + 5
```
Why?

kimishpatel replied:
I think this was out of habit for quantized stuff. I was trying to bias values of tensor rather than uniformly distributed between 0 and 1 to be able to catch error which are not easily canceled out. In this specific instance it may not matter.

dreiss commented on test/test_jit.py:
```diff
@@ -545,6 +545,153 @@ def f(x, y):
+        new_res = m(a, b)
+        FileCheck().check_not("aten::add(") \
+            .check_not("aten::relu_(") \
+            .check("aten::add_relu_(") \
```
Why is this `add_relu_`?  Shouldn't it be `add_relu`, since the original operand was not modified?  Doesn't the optimized model mutate `a`, whereas the original did not?

kimishpatel replied:
so add is not inplace but relu_ is. So it made sense to replace add + relu_ with add_relu_.

dreiss commented on test/test_jit.py:
```diff
@@ -545,6 +545,153 @@ def f(x, y):
+            .check_not("aten::relu(") \
+            .check("aten::add_relu_(") \
+            .run(m.graph)
+        torch.testing.assert_allclose(orig_res, new_res)
```
Should also test that both models properly mutate `a`.  Above, should probably test that they both *don't* mutate.

dreiss commented on torch/csrc/jit/passes/fuse_relu.cpp:
```diff
@@ -0,0 +1,94 @@
+        return (%res))";
+  // If add was followed by in place relu then
+  // we should probably make fused add_relu also
+  // inplace. since in the graph, it does not seem
+  // that output of add will be really used anywhere else.
```
I don't agree with this, since you're now mutating one of the addends, rather than the sum.

kimishpatel replied:
Hmm. I think you are right. I need to check a graph like this:
```
def m(a, b, c):
   tmp = a.add(b)
   d = a.add(c)
   x = self.relu_(tmp)
   return x + d
```
If add + relu_ is replaced with add_relu_ then a is mutated and if a.add(c) is executed afterwards that is wrong. This transformation adds a new alias. In the original graph `tmp = a.add(b)` a and tmp dont' alias but replacing that with `tmp = a.add_relu_(b)` now I have made a and tmp alias each other.
I will test and fix this.

kimishpatel replied:
@dreiss I fixed this and also added exactly this test that I mentioned above. Without fixing it, the test fails, after the fix it passes.

kimishpatel committed:
Update on "JIT pass for add relu fusion."


Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass for add relu fusion."


Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass for add relu fusion."


Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass for add relu fusion."


Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

[ghstack-poisoned]

kimishpatel committed:
Update on "JIT pass for add relu fusion."


Summary:
Building on top of previous PR that adds fused add_relu op, this PR adds
a JIT pass to transform input graph to find all fusable instancs of add
+ relu and fuses them.

Test Plan:
python test/test_jit.py TestJit.test_add_relu_fusion

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822396](https://our.internmc.facebook.com/intern/diff/D21822396)

[ghstack-poisoned]
